go/pkg/chunker/chunker.go

```diff
@@ -54,23 +60,79 @@ func NewFromBlob(blob []byte, chunkSize int) *Chunker {
 }
 
 // NewFromFile initializes a Chunker from the provided file.
-// The provided Digest has to match the contents of the file! If the size of the actual contents is
-// shorter than the provided Digest size, the Chunker will error on Next(), but otherwise the
-// results are unspecified.
+// The provided Digest does NOT have to match the contents of the file! It is
+// for informational purposes only. Chunker won't check Digest information at
+// any time. This means that late errors due to mismatch of digest and
+// contents are possible.
 func NewFromFile(path string, dg digest.Digest, chunkSize int) *Chunker {
+	c := NewFromReader(reader.NewFileReadSeeker(path, IOBufferSize), dg, chunkSize)
+	c.path = path
+	return c
+}
+
+// NewCompressedFromFile is similar to NewFromFile, except that it'll compress
+// the contents of the file using zstd compression.
+func NewCompressedFromFile(path string, dg digest.Digest, chunkSize int) (*Chunker, error) {
+	r, err := reader.NewCompressedFileSeeker(path, IOBufferSize)
+	if err != nil {
+		return nil, err
+	}
+	c := NewFromReader(r, dg, chunkSize)
+	c.path = path
+	return c, nil
+}
+
+// NewFromReader is similar to NewFromFile, except that it'll take in any
+// arbitrary reader.ReadSeeker. Note that the digest is still for convenience,
+// rather than being actual used by the Chunker.
+func NewFromReader(r reader.ReadSeeker, dg digest.Digest, chunkSize int) *Chunker {
 	if chunkSize < 1 {
 		chunkSize = DefaultChunkSize
 	}
 	if chunkSize > IOBufferSize {
 		chunkSize = IOBufferSize
 	}
 	return &Chunker{
+		r:         r,
 		chunkSize: chunkSize,
 		digest:    dg,
-		path:      path,
 	}
 }
 
+// CompressChunker will, if possible, modify the chunker so it'll return compressed
+// reads instead. This is irreversible.
+func CompressChunker(ch *Chunker) error {
+	var err error
+	if fullCompressor == nil {
+		fullCompressor, err = zstd.NewWriter(nil)
```

Comment:
This is racy, but it won't cause problems if you only use `EncodeAll`. But it would be better to just initialize it in global scope- as long as there are no invalid options specified, `zstd.NewWriter` doesn't return an error.

BTW- do you want to specify an explicit encoder level?